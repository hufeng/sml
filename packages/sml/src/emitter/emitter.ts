import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'

export abstract class Emitter<T> {
  protected sml: T
  protected s: Builder

  constructor(sml: T) {
    this.sml = sml
    this.s = new Builder()
  }

  abstract emitCode(): string

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
