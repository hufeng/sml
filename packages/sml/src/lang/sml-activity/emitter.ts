import S from '../s'
import { Emitter } from '../base'
import { Component, SmlActivityLangMeta } from './types'

export class PumlActivityEmitter extends Emitter<SmlActivityLangMeta> {
  emitCode() {
    const { components, links, vlinks, notes } = this.meta

    return this.s
      .$reset()
      .$s('@startuml')
      .$fn(this.buildTheme)
      .$fors(components, this.buildComponents)
      .$fors(links, this.buildLinks('->'))
      .$fors(vlinks, this.buildVlink('-->'))
      .$fors(notes, this.buildNotes)
      .$s('@enduml')
      .toString()
  }

  buildComponents(s: S, c: Component) {
    s.$s(`${c.type} "${c.label}" as ${c.id}`)
  }
}
