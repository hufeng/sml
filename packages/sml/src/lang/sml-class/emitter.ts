import { Emitter } from '../base'
import { ClassDiagramAst } from '../types'

export class PumlClassEmitter extends Emitter<ClassDiagramAst> {
  static visibleMap = {
    private: '-',
    protected: '#',
    package_private: '~',
    public: '+',
  } as const

  emitCode() {
    const { clazzes, structs, enums, interfaces, protocols } = this.meta
    this.emitStart()
      .emitInf(interfaces)
      .emitProtocol(protocols)
      .emitClazz(clazzes)
      .emitStruct(structs)
      .emitEnum(enums)
      .emitEnd()

    return this.s.toString('\n')
  }

  private emitStart() {
    this.s.$s('@startuml').$fn(this.buildConfig)
    return this
  }

  private emitEnd() {
    this.s.$s('@enduml')
    return this
  }

  /**
   *
   * @param s
   * @param clazzes
   */
  private emitClazz(clazzes: ClassDiagramAst['clazzes']) {
    this.s.$for(clazzes, (s, clazz) => {
      // extends and implements
      if (clazz.extends.length > 0 || clazz.implements.length > 0) {
        s.$s(
          `class ${clazz.name}${
            clazz.extends.length > 0
              ? ' extends ' + clazz.extends.join(', ')
              : ' '
          }${
            clazz.implements.length > 0
              ? ' implements ' + clazz.implements.join(', ')
              : ' '
          }`,
        )
      }

      // class start
      s.$s(`${clazz.abstract ? 'abstract' : ''} class ${clazz.name} {`)
      // fields
      s.$for(clazz.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(clazz.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })
    return this
  }

  /**
   * emit enums code
   * @param s
   * @param enums
   */
  private emitEnum(enums: ClassDiagramAst['enums']) {
    this.s.$for(enums, (s, e) => {
      s.$s(`enum ${e.name} {`)
        .$for(e.fields, (s, field) =>
          s.$s(
            `  ${field.name}${
              field.value !== 'undefined' ? ': ' + field.value : ''
            }`,
          ),
        )
        .$s('}')
    })

    return this
  }

  private emitInf(infs: ClassDiagramAst['interfaces']) {
    this.s.$for(infs, (s, i) => {
      // implements other interfaces
      if (i.implements.length > 0) {
        s.$s(`interface ${i.name} implements ${i.implements.join(', ')}`)
      }
      // start interface
      s.$s(`interface ${i.name} {`)
        .$for(i.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`  + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitProtocol(prots: ClassDiagramAst['protocols']) {
    this.s.$for(prots, (s, prot) => {
      // implements other interfaces
      if (prot.implements.length > 0) {
        s.$s(`interface ${prot.name} implements ${prot.implements.join(', ')}`)
      }
      s.$s(`protocol ${prot.name} {`)
        .$for(prot.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`   + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitStruct(structs: ClassDiagramAst['structs']) {
    this.s.$for(structs, (s, struct) => {
      // extends and implements
      if (struct.extends.length > 0 || struct.implements.length > 0) {
        s.$s(
          `class ${struct.name}${
            struct.extends.length > 0
              ? ' extends ' + struct.extends.join(', ')
              : ' '
          }${
            struct.implements.length > 0
              ? ' implements ' + struct.implements.join(', ')
              : ' '
          }`,
        )
      }
      // struct start
      s.$s(`struct ${struct.name} {`)
      // fields
      s.$for(struct.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(struct.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })

    return this
  }
}
