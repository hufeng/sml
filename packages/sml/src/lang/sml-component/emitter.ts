import { Emitter } from '../base'
import { SmlComponentAst } from '../types'

export class PumlComponentEmitter extends Emitter<SmlComponentAst> {
  emitCode() {
    const {
      packages,
      nodes,
      clouds,
      databases,
      components,
      infs,
      links,
      vlinks,
      rels,
    } = this.meta

    return this.s
      .str('@startuml')
      .forStr(
        packages,
        (s, v) =>
          s
            .str(`  package "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        packages.length > 0 ? '\n' : '',
      )
      .forStr(
        nodes,
        (s, v) =>
          s
            .str(`  node "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `   component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str(' }'),
        nodes.length > 0 ? '\n' : '',
      )
      .forStr(
        clouds,
        (s, v) =>
          s
            .str(`  cloud "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        clouds.length > 0 ? '\n' : '',
      )
      .forStr(
        databases,
        (s, v) =>
          s
            .str(`  database "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        databases.length > 0 ? '\n' : '',
      )
      .forStr(components, (s, v) =>
        s.str(`  component "${v.title}"${v.name ? '  as ' + v.name : ''}`),
      )
      .forStr(
        infs,
        (s, v) =>
          s.str(`  interface "${v.title}"${v.name ? '  as ' + v.name : ''}`),
        infs.length > 0 ? '\n' : '',
      )
      .forStr(links, (s, v) => s.str(`  ${v.from} --> ${v.to}`))
      .forStr(vlinks, (s, v) => s.str(`  ${v.from} ..> ${v.to}`))
      .forStr(rels, (s, v) => s.str(`  ${v.from} - ${v.to}`))
      .str('@enduml')
      .toString('\n')
  }
}
