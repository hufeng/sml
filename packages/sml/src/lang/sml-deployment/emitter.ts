import S from '../s'
import { Emitter } from '../base'
import { DeploymentLangAst, DeploymentBase, DeploymentContainer } from './types'

export class PumlDeploymentEmitter extends Emitter<DeploymentLangAst> {
  emitCode() {
    const { title, zones, components, links, vlinks, rels } = this.meta

    return this.s
      .reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$fn(this.buildConfig)
      .$s('')
      .$fors(components, this.build)
      .$for(zones, this.buildZones)
      .$for(links, this.buildLinks)
      .$for(vlinks, this.buildVLink)
      .$for(rels, this.buildRels)
      .$s('@enduml')
      .toString()
  }

  build(s: S, v: DeploymentBase) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`)
  }

  buildZones(s: S, v: DeploymentContainer) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>> {`)
      .$for(v.children, (s, v) =>
        s.$s(`  ${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`),
      )
      .$s('}')
  }
}
