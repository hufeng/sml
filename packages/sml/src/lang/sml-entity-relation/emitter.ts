import { Emitter } from '../base'
import S from '../s'
import { Entity, EntityField, EntityRelationDiagramAst, Relation } from './type'

export class PumlEntityRelationEmitter extends Emitter<EntityRelationDiagramAst> {
  constructor(ast: EntityRelationDiagramAst) {
    super(ast)
  }

  emitPuml(): string {
    const { title, entities, relations } = this.meta
    return this.s
      .$reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$s(this.buildTheme)
      .$s(this.buildTile)
      .$s(`hide circle`)
      .$s(`skinparam linetype ortho`)
      .$s('')
      .$for(entities, this.buildEntity)
      .$for(relations, this.buildRelation)
      .$s('@enduml')
      .toString()
  }

  private buildEntity(s: S, e: Entity) {
    return s
      .$s(`entity "${e.label}" as ${e.id} {`)
      .$for(e.ids, (s: S, f: EntityField) =>
        s.$s(`  *${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)).$s(`--`),
      )
      .$for(e.fields, (s: S, f: EntityField) =>
        s.$s(`  ${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)),
      )
      .$s('}')
  }

  private buildRelation(s: S, rel: Relation) {
    return s.$for(rel.to, (s: S, to) => s.$s(`${rel.from} ${rel.type} ${to}`))
  }
}
