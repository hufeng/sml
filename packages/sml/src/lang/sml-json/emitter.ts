import { Emitter } from '../base'
import { JsonDiagramAst } from './types'

export class PumlJsonEmitter extends Emitter<JsonDiagramAst> {
  emitCode() {
    const { title, highlights, json } = this.meta
    return this.s
      .reset()
      .$s(`@startjson ${title.replace(/ /g, '_')}`)
      .$fn(this.buildTheme)
      .$fors(highlights, (s, highlight) =>
        s.$s(
          `#highlight ${highlight
            .split('.')
            .map((str) => JSON.stringify(str))
            .join(' / ')}`,
        ),
      )
      .$s(JSON.stringify(json, null, 2))
      .$s('@endjson')
      .toString()
  }
}
