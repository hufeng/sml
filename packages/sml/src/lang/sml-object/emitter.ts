import { Emitter } from '../base'
import S from '../s'
import { ExtProp, MapNode, ObjectDiagramAst, ObjectNode } from './types'

export class SmlObjectEmitter extends Emitter<ObjectDiagramAst> {
  emitPuml() {
    const { title, obj, map, links, vlinks, rels, exts, compose, aggrate } = this.meta
    return this.s
      .$reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$s(this.buildDirection)
      .$s(this.buildTheme)
      .$s('')
      .$for(obj, this.buildObj)
      .$for(map, this.buildMap)
      .$s('')
      .$for(exts, (s: S, p: ExtProp) => s.$s(`${p.to} <|-- ${p.from}`))
      .$for(compose, (s: S, p: ExtProp) => s.$s(`${p.to} *-- ${p.from}`))
      .$for(aggrate, (s: S, p: ExtProp) => s.$s(`${p.to} o-- ${p.from}`))
      .$for(links, this.buildLinks())
      .$for(vlinks, this.buildLinks('..>'))
      .$for(rels, this.buildLinks('--'))
      .$s('@enduml')
      .toString()
  }

  emitMarkdown() {
    return ''
  }

  private buildObj(s: S, o: ObjectNode) {
    s.$s(`object "${o.label}" as ${o.id} {`)
      .$for(o.fields, (s, f) => s.$s(`  ${f.name} : ${f.val}`))
      .$s('}')
  }

  private buildMap(s: S, m: MapNode) {
    s.$s(`map "${m.label}" as ${m.id} {`)
      .$for(m.fields, (s, f) => s.$s(`  ${f.key} ${f.arrow} ${f.val}`))
      .$s('}')
  }
}
