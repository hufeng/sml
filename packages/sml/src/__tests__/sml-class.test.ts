import { PumlClassEmitter } from '../emitter/puml-class'
import * as sml from '../index'

describe('sml class diagram test suites', () => {
  const title = `suml test suite`

  it('test class', () => {
    const clazz = sml.ClassDiagram(title, (ml) => {
      ml.clazz(`org.hf.sml.UseCase`)
        .field('id', 'number')
        .field('name', 'string')
        .method('setId', ml.args(ml.arg('id', 'number')), ml.ret(ml.t.void))
        .method('getFullInof', ml.args(), ml.ret('string'))
      ml.struct(`org.hf.sml.Customer`)
        .field('id', 'number')
        .field('name', 'string')
        .method('setId', ml.args(ml.arg('id', 'number')), ml.ret(ml.t.void))
        .method('getFullInof', ml.args(), ml.ret('string'))
    })
    const code = new PumlClassEmitter(clazz).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test abstract class', () => {
    const abs = sml.ClassDiagram(title, (ml) => {
      ml.abstractClazz('or.hf.sml.AbstractUserCase')
        .field('id', 'number', 'private')
        .field('name', 'string')
        .method('setId', ml.args(ml.arg('id', 'number')), ml.ret(ml.t.void))
        .method('getFullInof', ml.args(), ml.ret('string'))
    })

    const code = new PumlClassEmitter(abs).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test enum', () => {
    const diagram = sml.ClassDiagram(title, (ml) => {
      ml.enum('org.hufeng.enum.Color')
        .field('Red', 0)
        .field('Green', 1)
        .field('Blue', 2)
      ml.enum('org.hufeng.enum.FileType')
        .field('Pdf', 'PDF')
        .field('doc', 'WORD')
        .field('xls', 'Excel')
    })

    const code = new PumlClassEmitter(diagram).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test inf', () => {
    const diagram = sml.ClassDiagram(title, (ml) => {
      ml.interface('org.hufeng.service.UserService')
        .method('sayYou', ml.args(ml.arg('name', 'string')))
        .method('sayMe', ml.args(ml.arg('you', 'string')))

      ml.protocol('org.hufeng.protocol.UserService')
        .method('sayYou', ml.args(ml.arg('name', 'string')))
        .method('sayMe', ml.args(ml.arg('you', 'string')))
    })

    const code = new PumlClassEmitter(diagram).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test extends and implements', () => {
    const diagram = sml.ClassDiagram(title, (ml) => {
      ml.interface(`List`)
      ml.abstractClazz('AbstractList')

      ml.clazz('org.hufeng.util.MyList')
        .extends('AbstractList')
        .implements('List')

      ml.interface('I1')
      ml.interface('I2')
      ml.interface('I3').implements('I1').implements('I2')
    })

    const code = new PumlClassEmitter(diagram).emitCode()
    expect(code).toMatchSnapshot()
  })
})
