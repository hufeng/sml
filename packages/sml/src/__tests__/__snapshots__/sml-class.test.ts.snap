// Vitest Snapshot v1

exports[`sml class diagram test suites > test abstract class 1`] = `
"@startuml
abstract class or.hf.sml.AbstractUserCase {
  - id: number;
  - name: string;

  + setId(id: number): void
  + getFullInof(): string
}

@enduml"
`;

exports[`sml class diagram test suites > test class 1`] = `
"@startuml
 class org.hf.sml.UseCase {
  - id: number;
  - name: string;

  + setId(id: number): void
  + getFullInof(): string
}

struct org.hf.sml.Customer {
  - id: number;
  - name: string;

  + setId(id: number): void
  + getFullInof(): string
}

@enduml"
`;

exports[`sml class diagram test suites > test enum 1`] = `
"@startuml
enum org.hufeng.enum.Color {
  Red: 0
  Green: 1
  Blue: 2
}
enum org.hufeng.enum.FileType {
  Pdf: PDF
  doc: WORD
  xls: Excel
}

@enduml"
`;

exports[`sml class diagram test suites > test extends and implements 1`] = `
"@startuml
interface List {
}
interface I1 {
}
interface I2 {
}
interface I3 implements I1, I2
interface I3 {
}

abstract class AbstractList {
}
class org.hufeng.util.MyList extends AbstractList implements List
 class org.hufeng.util.MyList {
}

@enduml"
`;

exports[`sml class diagram test suites > test inf 1`] = `
"@startuml
interface org.hufeng.service.UserService {
  + sayYou(name: string): void
  + sayMe(you: string): void
}

protocol org.hufeng.protocol.UserService {
   + sayYou(name: string): void
   + sayMe(you: string): void
}

@enduml"
`;
